controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +58,787 @@ ControllerManager::load_controller(
     [&](auto loader)
     {return loader->is_available(controller_type);});
 
-  std::shared_ptr<controller_interface::ControllerInterface> controller(nullptr);
+  controller_interface::ControllerInterfaceSharedPtr controller(nullptr);
   if (it != loaders_.cend()) {
     controller = (*it)->create(controller_type);
   } else {
     const std::string error_msg("Loader for controller '" + controller_name + "' not found\n");
     RCLCPP_ERROR(get_logger(), "%s", error_msg.c_str());
     throw std::runtime_error(error_msg);
   }
+  ControllerSpec controller_spec;
+  controller_spec.c = controller;
+  controller_spec.info.name = controller_name;
+  controller_spec.info.type = controller_type;
 
-  return add_controller_impl(controller, controller_name);
+  return add_controller_impl(controller_spec);
 }
 
-std::vector<std::shared_ptr<controller_interface::ControllerInterface>>
-ControllerManager::get_loaded_controllers() const
+controller_interface::ControllerInterfaceSharedPtr ControllerManager::load_controller(
+  const std::string & controller_name)
 {
-  return loaded_controllers_;
+  const std::string param_name = controller_name + ".type";
+  std::string controller_type;
+
+  // A priori we don't know the name of the controllers that will be loaded, so we cannot
+  // declare paramaters with their names.
+  // So when we're told to load a controller by name, we need to declare the parameter if
+  // we haven't done so, and then read it.
+
+  // Check if parameter has been declared
+  if (!has_parameter(param_name)) {
+    declare_parameter(param_name, rclcpp::ParameterValue());
+  }
+  if (!get_parameter(param_name, controller_type)) {
+    RCLCPP_ERROR(get_logger(), "'type' param not defined for %s", controller_name.c_str());
+    return nullptr;
+  }
+  return load_controller(controller_name, controller_type);
+}
+
+controller_interface::return_type ControllerManager::unload_controller(
+  const std::string & controller_name)
+{
+  std::lock_guard<std::recursive_mutex> guard(rt_controllers_wrapper_.controllers_lock_);
+  std::vector<ControllerSpec> & to = rt_controllers_wrapper_.get_unused_list(guard);
+  const std::vector<ControllerSpec> & from = rt_controllers_wrapper_.get_updated_list(guard);
+  to.clear();
```

Comment:
I don't really agree with the solution in this function.

The difference between `to` and `from` will be max one. So why not setting `to` equal to `from`? You can then easily call `std::find_if` on `to`, followed by `std::erase(std::remove(...))`.

Reply:
Agreed, again I expected that everyone would be more comfortable with keeping most of the old codebase and it would make the review process smoother.

I'll take a look at these functions and rewrite them.

Reply:
I believe I've rewritten and pushed all offending code.